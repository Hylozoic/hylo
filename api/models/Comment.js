import { compact, filter, sum } from 'lodash/fp'
import { markdown } from 'hylo-utils/text'
import decode from 'ent/decode'
import truncate from 'trunc-html'
import { parse } from 'url'

module.exports = bookshelf.Model.extend({
  tableName: 'comments',

  user: function () {
    return this.belongsTo(User)
  },

  post: function () {
    return this.belongsTo(Post)
  },

  text: function () {
    return this.get('text')
  },

  mentions: function () {
    return RichText.getUserMentions(this.text())
  },

  thanks: function () {
    return this.hasMany(Thank)
  },

  community: function () {
    return this.relations.post.relations.communities.first()
  },

  tags: function () {
    return this.belongsToMany(Tag).through(CommentTag)
  },

  activities: function () {
    return this.hasMany(Activity)
  },

  comment: function () {
    return this.belongsTo(Comment)
  },

  comments: function () {
    return this.hasMany(Comment, 'comment_id').query({where: {active: true}})
  },

  parentPost: function () {
    if (this.get('post_id')) {
      return this.load('post')
      .then(() => this.relations.post)
    } else if (this.get('comment_id')) {
      return this.load('comment')
      .then(() => this.relations.comment.parentPost())
    } else {
      return Promise.resolve()
    }
  },

  createActivities: function (trx) {    
    return this.load(['post', 'post.followers'])
    .then(() => {
      const followers = this.relations.post.relations.followers.map(follower => ({
        reader_id: follower.id,
        comment_id: this.id,
        post_id: this.relations.post.id,
        actor_id: this.get('user_id'),
        reason: 'newComment'
      }))
      const mentioned = RichText.getUserMentions(this.get('text')).map(mentionedId => ({
        reader_id: mentionedId,
        comment_id: this.id,
        post_id: this.relations.post.id,
        actor_id: this.get('user_id'),
        reason: 'commentMention'
      }))
      const readers = filter(r => r.reader_id !== this.get('user_id'), followers.concat(mentioned))
      return Activity.saveForReasons(readers, trx)
    })
  }
}, {

  find: function (id, options) {
    return Comment.where({id: id}).fetch(options)
  },

  createdInTimeRange: function (collection, startTime, endTime) {
    if (endTime === undefined) {
      endTime = startTime
      startTime = collection
      collection = Comment
    }

    return collection.query(function (qb) {
      qb.whereRaw('comments.created_at between ? and ?', [startTime, endTime])
      qb.where('comments.active', true)
    })
  },

  cleanEmailText: (user, text) => {
    text = text.replace(/\r\n/g, '\n').replace(/\r/g, '\n')
    const name = user.get('name').toLowerCase()
    const lines = text.split('\n')

    var cutoff
    lines.forEach((line, index) => {
      if (cutoff) return
      line = line.trim()

      if (line.length > 0 && name.startsWith(line.toLowerCase().replace(/^[\- ]*/, ''))) {
        // line contains only the user's name
        cutoff = index
        // also remove the common pattern of two dashes above the name
        if (index > 0 && lines[index - 1].match(/^\-\- ?$/)) {
          cutoff = index - 1
        }
      } else if (line.match(/^-{8}/)) {
        // line contains our divider, possibly followed by the text, "Only text
        // above the dashed line will be included."
        cutoff = index
      } else if (line.match(/(-{4,}.*dashed.line.*$)/)) {
        // line contains the divider at the end
        cutoff = index + 1
        lines[index] = line.replace(/(-{4,}.*dashed.line.*$)/, '')
      }
    })

    const finalText = cutoff ? lines.slice(0, cutoff).join('\n') : text
    return markdown(finalText || '')
  },

  notifyAboutMessage: ({commentId}) =>
    Comment.find(commentId, {withRelated: [
      'post.followers', 'post.lastReads'
    ]})
    .then(comment => {
      const { user_id, post_id, text } = comment.attributes
      const { followers, lastReads } = comment.relations.post.relations
      const recipients = followers.filter(u => u.id !== user_id)
      const user = followers.find(u => u.id === user_id)
      const alert = `${user.get('name')}: ${decode(truncate(text, 140).text).trim()}`
      const path = parse(Frontend.Route.thread({id: post_id})).path

      return Promise.map(recipients, user => {
        // don't notify if the user has read the thread recently.
        const lr = lastReads.find(r => r.get('user_id') === user.id)
        if (!lr || comment.get('created_at') > lr.get('last_read_at')) {
          return user.sendPushNotification(alert, path)
        }
      })
    }),

  sendMessageDigests: (minutes = 10) => {
    const time = new Date(new Date() - minutes * 60000)

    return Post.where('type', Post.Type.THREAD)
    .where('updated_at', '>', time)
    .fetchAll({withRelated: [
      'followers',
      'followers.devices',
      'lastReads',
      {comments: q => {
        q.where('created_at', '>', time)
        q.orderBy('created_at', 'asc')
      }},
      'comments.user'
    ]})
    .then(posts => Promise.all(posts.map(post => {
      const { comments, followers, lastReads } = post.relations
      if (comments.length === 0) return

      return Promise.map(followers.models, user => {
        // don't send to users that have devices (they receive push
        // notifications instead)
        if (user.relations.devices.length > 0) return

        // select comments not written by this user and newer than user's last
        // read time
        const r = lastReads.find(l => l.get('user_id') === user.id)
        const filtered = comments.filter(c =>
          c.get('created_at') > (r ? r.get('last_read_at') : 0) &&
          c.get('user_id') !== user.id)
        if (filtered.length === 0) return

        // here, we assume that all of the messages were sent by 1 other person,
        // so this will have to change when we support group messaging
        const other = filtered[0].relations.user

        return Email.sendMessageDigest({
          email: user.get('email'),
          data: {
            other_person_avatar_url: other.get('avatar_url'),
            other_person_name: other.get('name'),
            thread_url: Frontend.Route.thread(post),
            messages: filtered.map(c => c.get('text'))
          }
        })
      })
      .then(sends => compact(sends).length)
    })))
    .then(postSendCounts => sum(postSendCounts))
  }
})
