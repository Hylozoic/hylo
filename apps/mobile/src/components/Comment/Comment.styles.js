import { StyleSheet } from 'react-native'
import { limedSpruce } from '@hylo/presenters/colors'
import Colors from '../../style/theme-colors'

export default {
  container: {
    flexDirection: 'column',
    flex: 1,
    padding: 10,
    paddingTop: 3,
    paddingBottom: 3,
    paddingRight: 10
  },
  avatar: {
    marginRight: 10
  },
  header: {
    flexDirection: 'row',
    justifyItems: 'stretch',
    marginBottom: 3
  },
  headerLeft: {
  },
  headerMiddle: {
    flex: 1
  },
  headerRight: {
    paddingTop: 3,
    gap: 10,
    flexDirection: 'row',
    alignItems: 'flex-start'
  },
  highlighted: {
    backgroundColor: 'rgb(255,255,224)'
  },
  menuIcon: {
    fontSize: 20,
    paddingLeft: 10,
    color: Colors.foreground50
  },
  actionSheetIcon: {
    fontSize: 20
  },
  name: {
    color: limedSpruce,
    fontFamily: 'Circular-Bold',
    fontSize: 14
  },
  date: {
    color: Colors.mutedForeground,
    fontFamily: 'Circular-Book',
    fontSize: 12
  },
  replyLink: {
    flexDirection: 'row'
  },
  // replyLinkText: {
  //   fontSize: 12,
  //   color: slateGrey80,
  //   marginLeft: 5
  // },
  replyLinkIcon: {
    transform: [{ rotateY: '180deg' }],
    fontSize: 20,
    color: Colors.mutedForeground
  },
  imageAttachment: {
    borderRadius: 10
  },
  body: {
    paddingLeft: 42
  }
}
