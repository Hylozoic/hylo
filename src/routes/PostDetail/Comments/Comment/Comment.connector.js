import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { isEmpty } from 'lodash/fp'
import getGroupForCurrentRoute from 'store/selectors/getGroupForCurrentRoute'
import createComment from 'store/actions/createComment'
import updateComment from 'store/actions/updateComment'
import deleteComment from 'store/actions/deleteComment'
import fetchChildComments from 'store/actions/fetchChildComments'
import { getHasMoreChildComments, getTotalChildComments } from 'store/selectors/getChildComments'
import getMe from 'store/selectors/getMe'

export function mapStateToProps (state, props) {
  const { comment } = props
  const currentUser = getMe(state, props)
  const group = getGroupForCurrentRoute(state, props)
  const isCreator = currentUser && (comment.creator.id === currentUser.id)
  const canModerate = currentUser && currentUser.canModerate(group)
  console.log("comment connector!", comment)
  return {
    childCommentsTotal: getTotalChildComments(state, { id: comment.id }),
    hasMoreChildComments: getHasMoreChildComments(state, { id: comment.id }),
    canModerate,
    isCreator
  }
}

export const mapDispatchToProps = (dispatch, props) => {
  const { postId, comment } = props
  return {
    ...bindActionCreators({
      deleteComment,
      updateComment
    }, dispatch),
    fetchCommentsMaker: cursor => () => dispatch(fetchChildComments(comment.id, { cursor })),
    createComment: commentParams => dispatch(createComment({
      postId,
      parentCommentId: comment.id,
      ...commentParams
    }))
  }
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { canModerate, isCreator } = stateProps
  const { fetchCommentsMaker } = dispatchProps
  const { comment } = ownProps

  const deleteCommentWithConfirm = isCreator
    ? (commentId) => window.confirm('Are you sure you want to delete this comment?') &&
      dispatchProps.deleteComment(commentId)
    : null

  const removeCommentWithConfirm = !isCreator && canModerate
    ? (commentId) => window.confirm('Are you sure you want to remove this comment?') &&
      dispatchProps.deleteComment(commentId)
    : null

  const updateComment = isCreator
    ? (commentId, text) => dispatchProps.updateComment(commentId, text)
    : null

  const cursor = !isEmpty(comment.childComments) && comment.childComments[0].id
  const fetchChildComments = fetchCommentsMaker(cursor)

  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    deleteComment: deleteCommentWithConfirm,
    removeComment: removeCommentWithConfirm,
    updateComment,
    fetchChildComments
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
