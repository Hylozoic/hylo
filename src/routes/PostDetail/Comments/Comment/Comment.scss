$comment-left-margin: 69px;

.comment {
  padding: 0px 25px;
  margin-bottom: $space-6x;
 }

.header {
  display: flex;
  align-items: center;
}

.avatar {
  margin-right: 25px;
}

.userName {
  composes: userName from 'components/PostCard/PostHeader/PostHeader.scss';
  margin-right: 7px;
}

.commented {
  composes: caption-lt-lg from 'css/typography.scss';
}

.upperRight {
  margin-left: auto;
}

.timestamp {
  composes: timestamp from 'css/typography.scss';
  margin-right: 10px;
  position: relative;
  top: -3px;
}

.dropdown {
/* !empty */
}

.text {
  composes: bdy-lt-sm from 'css/typography.scss';
  margin-left: $comment-left-margin;
}

.reply {
  composes: text-button from 'css/typography.scss';
  margin-bottom: 20px;
  &:hover {
    //hack to get the icon to change on hover
    .icon {
      color: $color-gossamer;
    }
  }
}

.imageLink {
  display: block;
  margin-left: 69px;
}

.images {
  display: flex;
  padding-top: 0px;
  margin-left: $comment-left-margin;
  width: 70%;
  border-radius: 20px;
}

.files {
  display: flex;
  margin-left: $comment-left-margin
}

.editor {
  composes: bdy-lt-sm from 'css/typography.scss';
  margin-left: $comment-left-margin;
  padding: 3px 5px 10px 5px;
  border: 1px solid $color-cape-cod-10;
}


@media screen and (max-width: 425px) {
  .comment {
    padding: 0 10px;
  }

  .avatar {
    margin-right: 10px;
  }

  .text {
    margin-left: 54px;
  }
}
