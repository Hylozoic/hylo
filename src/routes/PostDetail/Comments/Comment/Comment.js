import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import ReactTooltip from 'react-tooltip'
import { filter, isEmpty, isFunction, reverse } from 'lodash/fp'
import { humanDate, present, sanitize } from 'hylo-utils/text'
import { personUrl } from 'util/navigation'
import Avatar from 'components/Avatar'
import Dropdown from 'components/Dropdown'
import Icon from 'components/Icon'
import ClickCatcher from 'components/ClickCatcher'
import HyloEditor from 'components/HyloEditor'
import contentStateToHTML from 'components/HyloEditor/contentStateToHTML'
import CardImageAttachments from 'components/CardImageAttachments'
import CardFileAttachments from 'components/CardFileAttachments'
import CommentForm from '../CommentForm'
import './Comment.scss'

const { object, func } = PropTypes

class Comment extends Component {
  static propTypes = {
    comment: object.isRequired,
    onReplyComment: func.isRequired
  }

  state = {
    editing: false
  }

  editComment = () => {
    this.setState({ editing: true })
  }

  saveComment = editorState => {
    const { comment } = this.props
    const contentState = editorState.getCurrentContent()
    if ((!contentState.hasText() || isEmpty(contentState.getPlainText().trim())) && isEmpty(comment.attachments)) {
      // Don't accept empty comments.
      return
    }

    this.setState({ editing: false })
    this.props.updateComment(contentStateToHTML(editorState.getCurrentContent()))
  }

  render () {
    const { comment, slug, isCreator, onReplyComment, deleteComment, removeComment } = this.props
    const { id, creator, createdAt, text, attachments } = comment
    const { editing } = this.state
    const profileUrl = personUrl(creator.id, slug)
    const presentedText = present(sanitize(text), { slug })

    const dropdownItems = filter(item => isFunction(item.onClick), [
      {},
      { icon: 'Edit', label: 'Edit', onClick: isCreator && this.editComment },
      { icon: 'Trash', label: 'Delete', onClick: deleteComment },
      { icon: 'Trash', label: 'Remove', onClick: removeComment }
    ])

    return (
      <div>
        <div styleName='header'>
          <Avatar avatarUrl={creator.avatarUrl} url={profileUrl} styleName='avatar' />
          <Link to={profileUrl} styleName='userName'>{creator.name}</Link>
          <span styleName='timestamp'>
            {editing && 'Editing now'}
            {!editing && humanDate(createdAt)}
          </span>
          <div styleName='upperRight'>
            <div styleName='commentAction' onClick={onReplyComment} data-tip='Reply' data-for={`reply-tip-${id}`}>
              <Icon name='Replies' />
            </div>
            {dropdownItems.length > 0 && <Dropdown styleName='dropdown' toggleChildren={<Icon name='More' />} items={dropdownItems} />}
          </div>
        </div>
        <CardImageAttachments attachments={attachments} linked styleName='images' />
        <CardFileAttachments attachments={attachments} styleName='files' />
        <ClickCatcher>
          {editing && <HyloEditor
            styleName='editor'
            onChange={this.startTyping}
            contentHTML={text}
            parentComponent={'CommentForm'}
            submitOnReturnHandler={this.saveComment} />}
          {!editing && <div id='text' styleName='text' dangerouslySetInnerHTML={{ __html: presentedText }} />}
        </ClickCatcher>
      </div>
    )
  }
}

export default class CommentWithReplies extends Component {
  static propTypes = {
    comment: object.isRequired,
    createComment: func.isRequired // bound by Comments.connector & Comment.connector
  }

  static defaultProps = {
    attachments: []
  }

  state = {
    replying: false,
    prefillEditor: null
  }

  onReplyComment = (e, toMemberSlug) => {
    // On any interaction, relevant comment box shows & only leaves
    // naturally once the component is cleared from view.
    this.setState({
      replying: true,
      prefillEditor: toMemberSlug ? `<p>@${toMemberSlug} </p>` : ''
    })
  }

  render () {
    const { comment, createComment } = this.props
    const { childComments } = comment
    const { replying } = this.state

    return <div styleName='comment'>
      <Comment {...this.props} onReplyComment={this.onReplyComment} />
      {childComments && childComments.items && <div styleName='subreply'>
        {reverse(childComments.items).map(c =>
          <Comment key={c.id}
            {...this.props}
            comment={c}
            // sets child comments to toggle reply box one level deep, rather than allowing recursion
            onReplyComment={(e) => this.onReplyComment(e, c.creator.name)}
          />)
        }
      </div>}
      {replying && <div styleName='replybox'>
        <CommentForm
          createComment={createComment}
          placeholder={`Reply to ${comment.creator.name}`}
          editorContent={this.state.prefillEditor}
          focusOnRender />
      </div>}
      <ReactTooltip
        id={`reply-tip-${comment.id}`}
        effect='solid'
        style='light'
        border
        // :TODO: de-duplicate these colour values
        textColor='#2A4059'
        borderColor='#40A1DD'
        backgroundColor='white'
        offset={{ 'top': -2 }}
        delayShow={500}
        styleName='actionsTooltip' />
    </div>
  }
}
