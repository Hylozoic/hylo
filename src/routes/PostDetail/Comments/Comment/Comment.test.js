import { Comment } from './Comment'
import { shallow } from 'enzyme'
import React from 'react'

jest.mock('components/HyloEditor/contentStateToHTML', () => any => any)

describe('Comment', () => {
  const props = {
    comment: {
      text: '<p>text of the comment</p>',
      creator: {
        id: 1,
        name: 'Joe Smith',
        avatarUrl: 'foo.jpg'
      },
      attachments: [],
      createdAt: new Date(),
      childComments: []
    },
    slug: 'foo',
    updateComment: jest.fn(),
    deleteComment: jest.fn(),
    removeComment: jest.fn(),
    onReplyComment: jest.fn()
  }

  it('renders correctly', () => {
    const wrapper = shallow(<Comment {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly when editing', () => {
    const wrapper = shallow(<Comment {...props} />)
    wrapper.instance().setState({ editing: true })
    wrapper.update()
    expect(wrapper).toMatchSnapshot()
  })

  it('displays image attachments', () => {
    const comment = {
      ...props.comment,
      attachments: [
        { url: 'foo.png', attachmentType: 'image' }
      ]
    }
    const wrapper = shallow(<Comment {...props} comment={comment} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('sanitizes text', () => {
    const comment = {
      ...props.comment,
      text: '<p>Nice text<script>a sneaky script</script></p>'
    }
    const wrapper = shallow(<Comment {...props} comment={comment} />)
    expect(wrapper.find('div #text').prop('dangerouslySetInnerHTML')).toEqual({
      __html: '<p>Nice text</p>'
    })
  })

  it('displays the delete menu when deleteComment is defined', () => {
    const wrapper = shallow(<Comment {...props} deleteComment={() => {}} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('displays the remove menu when removeComment is defined', () => {
    const wrapper = shallow(<Comment {...props} removeComment={() => {}} />)
    expect(wrapper).toMatchSnapshot()
  })

  describe('editComment', () => {
    it('sets state.editing to true', () => {
      const wrapper = shallow(<Comment {...props} />)
      const instance = wrapper.instance()
      expect(instance.state.editing).toEqual(false)
      instance.editComment()
      expect(instance.state.editing).toEqual(true)
    })
  })

  describe('saveComment', () => {
    it('sets state.editing to false and calls props.updateComment', () => {
      const wrapper = shallow(<Comment {...props} />)
      const theText = 'lalala'
      const editorState = {
        getCurrentContent: () => ({
          hasText: () => true,
          getPlainText: () => theText
        })
      }
      const instance = wrapper.instance()
      instance.setState({ editing: true })
      instance.saveComment(editorState)
      expect(instance.state.editing).toEqual(false)
      expect(props.updateComment).toHaveBeenCalled()
    })
  })
})
