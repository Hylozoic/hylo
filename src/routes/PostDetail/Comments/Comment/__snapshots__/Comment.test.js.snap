// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment displays image attachments 1`] = `
<div>
  <div
    data-stylename="header"
  >
    <Avatar
      avatarUrl="foo.jpg"
      data-stylename="avatar"
      url="/groups/foo/members/1"
    />
    <Link
      data-stylename="userName"
      to="/groups/foo/members/1"
    >
      Joe Smith
    </Link>
    <span
      data-stylename="timestamp"
    >
      just now
    </span>
    <div
      data-stylename="upperRight"
    >
      <div
        data-for="reply-tip-undefined"
        data-stylename="commentAction"
        data-tip="Reply"
        onClick={[MockFunction]}
      >
        <Icon
          name="Replies"
        />
      </div>
    </div>
  </div>
  <CardImageAttachments
    attachments={
      Array [
        Object {
          "attachmentType": "image",
          "url": "foo.png",
        },
      ]
    }
    data-stylename="images"
    linked={true}
  />
  <CardFileAttachments
    attachments={
      Array [
        Object {
          "attachmentType": "image",
          "url": "foo.png",
        },
      ]
    }
    data-stylename="files"
  />
  <ClickCatcher
    groupSlug="foo"
  >
    <ForwardRef(HyloEditor)
      contentHTML="<p>text of the comment</p>"
      data-stylename="text"
      onEnter={[Function]}
      onEscape={[Function]}
      readOnly={true}
    />
    <EmojiRow
      attachments={
        Array [
          Object {
            "attachmentType": "image",
            "url": "foo.png",
          },
        ]
      }
      childComments={Array []}
      createdAt={2022-10-30T21:11:38.942Z}
      creator={
        Object {
          "avatarUrl": "foo.jpg",
          "id": 1,
          "name": "Joe Smith",
        }
      }
      currentUser={
        Object {
          "id": 2,
        }
      }
      text="<p>text of the comment</p>"
    />
  </ClickCatcher>
</div>
`;

exports[`Comment displays the delete menu when deleteComment is defined 1`] = `
<div>
  <div
    data-stylename="header"
  >
    <Avatar
      avatarUrl="foo.jpg"
      data-stylename="avatar"
      url="/groups/foo/members/1"
    />
    <Link
      data-stylename="userName"
      to="/groups/foo/members/1"
    >
      Joe Smith
    </Link>
    <span
      data-stylename="timestamp"
    >
      just now
    </span>
    <div
      data-stylename="upperRight"
    >
      <div
        data-for="reply-tip-undefined"
        data-stylename="commentAction"
        data-tip="Reply"
        onClick={[MockFunction]}
      >
        <Icon
          name="Replies"
        />
      </div>
      <Dropdown
        data-stylename="dropdown"
        items={
          Array [
            Object {
              "icon": "Edit",
              "label": "Edit",
              "onClick": [Function],
            },
            Object {
              "icon": "Trash",
              "label": "Delete",
              "onClick": [Function],
            },
          ]
        }
        toggleChildren={
          <Icon
            name="More"
          />
        }
      />
    </div>
  </div>
  <CardImageAttachments
    attachments={Array []}
    data-stylename="images"
    linked={true}
  />
  <CardFileAttachments
    attachments={Array []}
    data-stylename="files"
  />
  <ClickCatcher
    groupSlug="foo"
  >
    <ForwardRef(HyloEditor)
      contentHTML="<p>text of the comment</p>"
      data-stylename="text"
      onEnter={[Function]}
      onEscape={[Function]}
      readOnly={true}
    />
    <EmojiRow
      attachments={Array []}
      childComments={Array []}
      createdAt={2022-10-30T21:11:38.942Z}
      creator={
        Object {
          "avatarUrl": "foo.jpg",
          "id": 1,
          "name": "Joe Smith",
        }
      }
      currentUser={
        Object {
          "id": 1,
        }
      }
      text="<p>text of the comment</p>"
    />
  </ClickCatcher>
</div>
`;

exports[`Comment displays the remove menu when removeComment is defined 1`] = `
<div>
  <div
    data-stylename="header"
  >
    <Avatar
      avatarUrl="foo.jpg"
      data-stylename="avatar"
      url="/groups/foo/members/1"
    />
    <Link
      data-stylename="userName"
      to="/groups/foo/members/1"
    >
      Joe Smith
    </Link>
    <span
      data-stylename="timestamp"
    >
      just now
    </span>
    <div
      data-stylename="upperRight"
    >
      <div
        data-for="reply-tip-undefined"
        data-stylename="commentAction"
        data-tip="Reply"
        onClick={[MockFunction]}
      >
        <Icon
          name="Replies"
        />
      </div>
      <Dropdown
        data-stylename="dropdown"
        items={
          Array [
            Object {
              "icon": "Trash",
              "label": "Remove",
              "onClick": [Function],
            },
          ]
        }
        toggleChildren={
          <Icon
            name="More"
          />
        }
      />
    </div>
  </div>
  <CardImageAttachments
    attachments={Array []}
    data-stylename="images"
    linked={true}
  />
  <CardFileAttachments
    attachments={Array []}
    data-stylename="files"
  />
  <ClickCatcher
    groupSlug="foo"
  >
    <ForwardRef(HyloEditor)
      contentHTML="<p>text of the comment</p>"
      data-stylename="text"
      onEnter={[Function]}
      onEscape={[Function]}
      readOnly={true}
    />
    <EmojiRow
      attachments={Array []}
      childComments={Array []}
      createdAt={2022-10-30T21:11:38.942Z}
      creator={
        Object {
          "avatarUrl": "foo.jpg",
          "id": 1,
          "name": "Joe Smith",
        }
      }
      currentUser={
        Object {
          "id": 2,
        }
      }
      text="<p>text of the comment</p>"
    />
  </ClickCatcher>
</div>
`;

exports[`Comment does not display the delete menu when deleteComment is not defined 1`] = `
<div>
  <div
    data-stylename="header"
  >
    <Avatar
      avatarUrl="foo.jpg"
      data-stylename="avatar"
      url="/groups/foo/members/1"
    />
    <Link
      data-stylename="userName"
      to="/groups/foo/members/1"
    >
      Joe Smith
    </Link>
    <span
      data-stylename="timestamp"
    >
      just now
    </span>
    <div
      data-stylename="upperRight"
    >
      <div
        data-for="reply-tip-undefined"
        data-stylename="commentAction"
        data-tip="Reply"
        onClick={[MockFunction]}
      >
        <Icon
          name="Replies"
        />
      </div>
    </div>
  </div>
  <CardImageAttachments
    attachments={Array []}
    data-stylename="images"
    linked={true}
  />
  <CardFileAttachments
    attachments={Array []}
    data-stylename="files"
  />
  <ClickCatcher
    groupSlug="foo"
  >
    <ForwardRef(HyloEditor)
      contentHTML="<p>text of the comment</p>"
      data-stylename="text"
      onEnter={[Function]}
      onEscape={[Function]}
      readOnly={true}
    />
    <EmojiRow
      attachments={Array []}
      childComments={Array []}
      createdAt={2022-10-30T21:11:38.942Z}
      creator={
        Object {
          "avatarUrl": "foo.jpg",
          "id": 1,
          "name": "Joe Smith",
        }
      }
      currentUser={
        Object {
          "id": 2,
        }
      }
      text="<p>text of the comment</p>"
    />
  </ClickCatcher>
</div>
`;

exports[`Comment does not display the remove menu when removeComment is not defined 1`] = `
<div>
  <div
    data-stylename="header"
  >
    <Avatar
      avatarUrl="foo.jpg"
      data-stylename="avatar"
      url="/groups/foo/members/1"
    />
    <Link
      data-stylename="userName"
      to="/groups/foo/members/1"
    >
      Joe Smith
    </Link>
    <span
      data-stylename="timestamp"
    >
      just now
    </span>
    <div
      data-stylename="upperRight"
    >
      <div
        data-for="reply-tip-undefined"
        data-stylename="commentAction"
        data-tip="Reply"
        onClick={[MockFunction]}
      >
        <Icon
          name="Replies"
        />
      </div>
      <Dropdown
        data-stylename="dropdown"
        items={
          Array [
            Object {
              "icon": "Edit",
              "label": "Edit",
              "onClick": [Function],
            },
            Object {
              "icon": "Trash",
              "label": "Delete",
              "onClick": [Function],
            },
          ]
        }
        toggleChildren={
          <Icon
            name="More"
          />
        }
      />
    </div>
  </div>
  <CardImageAttachments
    attachments={Array []}
    data-stylename="images"
    linked={true}
  />
  <CardFileAttachments
    attachments={Array []}
    data-stylename="files"
  />
  <ClickCatcher
    groupSlug="foo"
  >
    <ForwardRef(HyloEditor)
      contentHTML="<p>text of the comment</p>"
      data-stylename="text"
      onEnter={[Function]}
      onEscape={[Function]}
      readOnly={true}
    />
    <EmojiRow
      attachments={Array []}
      childComments={Array []}
      createdAt={2022-10-30T21:11:38.942Z}
      creator={
        Object {
          "avatarUrl": "foo.jpg",
          "id": 1,
          "name": "Joe Smith",
        }
      }
      currentUser={
        Object {
          "id": 1,
        }
      }
      text="<p>text of the comment</p>"
    />
  </ClickCatcher>
</div>
`;

exports[`Comment renders correctly 1`] = `
<div>
  <div
    data-stylename="header"
  >
    <Avatar
      avatarUrl="foo.jpg"
      data-stylename="avatar"
      url="/groups/foo/members/1"
    />
    <Link
      data-stylename="userName"
      to="/groups/foo/members/1"
    >
      Joe Smith
    </Link>
    <span
      data-stylename="timestamp"
    >
      just now
    </span>
    <div
      data-stylename="upperRight"
    >
      <div
        data-for="reply-tip-undefined"
        data-stylename="commentAction"
        data-tip="Reply"
        onClick={[MockFunction]}
      >
        <Icon
          name="Replies"
        />
      </div>
    </div>
  </div>
  <CardImageAttachments
    attachments={Array []}
    data-stylename="images"
    linked={true}
  />
  <CardFileAttachments
    attachments={Array []}
    data-stylename="files"
  />
  <ClickCatcher
    groupSlug="foo"
  >
    <ForwardRef(HyloEditor)
      contentHTML="<p>text of the comment</p>"
      data-stylename="text"
      onEnter={[Function]}
      onEscape={[Function]}
      readOnly={true}
    />
    <EmojiRow
      attachments={Array []}
      childComments={Array []}
      createdAt={2022-10-30T21:11:38.942Z}
      creator={
        Object {
          "avatarUrl": "foo.jpg",
          "id": 1,
          "name": "Joe Smith",
        }
      }
      currentUser={
        Object {
          "id": 2,
        }
      }
      text="<p>text of the comment</p>"
    />
  </ClickCatcher>
</div>
`;

exports[`Comment renders correctly when editing 1`] = `
<div>
  <div
    data-stylename="header"
  >
    <Avatar
      avatarUrl="foo.jpg"
      data-stylename="avatar"
      url="/groups/foo/members/1"
    />
    <Link
      data-stylename="userName"
      to="/groups/foo/members/1"
    >
      Joe Smith
    </Link>
    <span
      data-stylename="timestamp"
    >
      Editing now
    </span>
    <div
      data-stylename="upperRight"
    >
      <div
        data-for="reply-tip-undefined"
        data-stylename="commentAction"
        data-tip="Reply"
        onClick={[MockFunction]}
      >
        <Icon
          name="Replies"
        />
      </div>
      <Icon
        data-stylename="cancelIcon"
        name="Ex"
        onClick={[Function]}
      />
    </div>
  </div>
  <CardImageAttachments
    attachments={Array []}
    data-stylename="images"
    linked={true}
  />
  <CardFileAttachments
    attachments={Array []}
    data-stylename="files"
  />
  <ClickCatcher
    groupSlug="foo"
  >
    <ForwardRef(HyloEditor)
      contentHTML="<p>text of the comment</p>"
      data-stylename="editing"
      onEnter={[Function]}
      onEscape={[Function]}
      readOnly={false}
    />
    <EmojiRow
      attachments={Array []}
      childComments={Array []}
      createdAt={2022-10-30T21:11:38.942Z}
      creator={
        Object {
          "avatarUrl": "foo.jpg",
          "id": 1,
          "name": "Joe Smith",
        }
      }
      currentUser={
        Object {
          "id": 2,
        }
      }
      text="<p>text of the comment</p>"
    />
  </ClickCatcher>
</div>
`;
